### ISSUE TITLE
AllRecordsRange misses first record

### ISSUE THREAD

**Issue by jblachly** [blachlylab/dhtslib#59]
First record somehow gets burned and never returned wiht `foreach`

Demonstrated with an integration test, push incoming

**Comment by jblachly** [comment]
Is this the problem?

https://github.com/blachlylab/dhtslib/blob/428266a6b5516493a9a9770be689b260d80a67f9/source/dhtslib/sam/package.d#L1092

If so I am having a hard time following why

**Comment by jblachly** [comment]
Yep, that's the problem. However, fixing it makes 3 of 8 unittests fail ...

**Comment by jblachly** [comment]
This may also be a samtools `sam_read1` bug ...

```
[I::test.samreader.main] Test AllRecordsRange iterator
empty
front

popFront
empty
front
HS18_09653:4:1308:11522:27107
popFront
empty
front
HS18_09653:4:2314:14991:85680
```

**Comment by jblachly** [comment]
So the use of SAMFile in the test script prior to AllRecordsRange to do queries causes the first call to sam_read1 to return the second record ...

**Comment by charlesgregory** [comment]
Might need to call an htslib function to make sure file iteration is reset when calling allRecords.

On Mar 8, 2021, at 9:04 PM, James S Blachly, MD <notifications@github.com> wrote:

﻿

So the use of SAMFile in the test script prior to AllRecordsRange to do queries causes the first call to sam_read1 to return the second record ...

**Comment by jblachly** [comment]
Looking for it now, but doesn't appear to be antyhing along the lines of "seek" etc.

**Comment by jblachly** [comment]
@charlesgregory  found it.

It is actually `RecordRange` with naughty behavior (calling the iterator upon range initialization in the ctor without first checking that the iterator has results) coupled with the immediately preceding test being a test for query with zero results. The extra call to `sam_itr_next` when `itr.finished` is actually true, but hasn't been checked, consumes the first read of the bamfile and advances the fp.

https://github.com/blachlylab/dhtslib/blob/428266a6b5516493a9a9770be689b260d80a67f9/source/dhtslib/sam/package.d#L1154-L1163

i'm fixing this in a series of commits

**Comment by jblachly** [comment]
Further note to self, in fact the last read of a query iterator always consumes an read (apparently) when it returns -1 (i.e., consumes a read that's not returned in the results set), and `sam_read1` always picks up after where the iterator left off.

will need probably to directly manipulate the fp to `seek(0)` in case of `AllRecordsRange`
